### PR TITLE
MGDAPI-5220 Updates to GCP E2E tests

### PR DESCRIPTION
**Pull request by Fiona-Waters**
# Issue link
[MGDAPI-5220](https://issues.redhat.com/browse/MGDAPI-5220)
# What
Making some updates to GCP E2E tests as well as refactoring to allow for both cloud providers.

# Verification steps
Provision a GCP CCS Cluster - request an access key and place it in the root of the delorean repo.
Run: 
```
OCM_CLUSTER_NAME=<yourname>-ccs OCM_CLUSTER_LIFESPAN=24 COMPUTE_NODES_COUNT=4 BYOC=true CLOUD_PROVIDER=gcp make -f make/ocm.mk ocm/cluster.json
make -f make/ocm.mk ocm/cluster/create
```

Once the cluster is ready:
- Checkout this pr `gh pr checkout 3117`
- Install RHOAM via OLM
  - build and push the RHOAM binary
```
ORG=<Your Quay Org> OLM_TYPE=managed-api-service INSTALLATION _TYPE=managed-api make image/build/push
```
  - Edit the managed-api-service.clusterserviceversion.yaml to refer to your image 
  - Build bundle and index images
```
ORG=<Your Quay Org> OLM_TYPE=managed-api-service UPGRADE=false BUNDLE_VERSIONS=1.31.0 make create/olm/bundle
```
 - Login to the OSD UI and Create a catalog source that points to your index image.
 ```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: rhmi-operators
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/<your org>/<OLM_TYPE>-index:<index version>
```
- Prepare the cluster
```
INSTALLATION_TYPE=managed-api LOCAL=false make cluster/prepare/local
```
- Create RHMI CR
```
LOCAL=false INSTALLATION_TYPE=managed-api make deploy/integreatly-rhmi-cr.yml
```
- Install RHOAM from OSD operator hub

- Once RHOAM is installed build and run the E2E tests on cluster.
```
ORG=<Your Quay Org> make image/functional/build
ORG=<Your Quay Org> make image/functional/push
```

- Create a `test-config.yaml` file in the root directory containing the following, adding your test image url:
```
tests:
  - name: "integreatly-operator-test"
    image: <"Your Test Image">
    timeout: 7200
    envVars:
      - name: "DESTRUCTIVE"
        value: "false"
      - name: "MULTIAZ"
        value: "false"
      - name: "WATCH_NAMESPACE"
        value: "redhat-rhoam-operator"
```
- Create another file called `testOutput.txt` and run the following command:
```
KUBECONFIG=~/.kube/config DESTRUCTIVE=false MULTIAZ=false./delorean pipeline product-tests --test-config ./test-config.yaml --output test-results --namespace test-functional | tee testOutput.txt

```
- Check that all tests pass.

- Uninstall RHOAM and cleanup/delete cluster

### CODE REVIEW

File: test/functional/gcp_network_state.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+	// get the cidr block from Strategy Map
+	expectedCidr, err := verifyAndGetCidrBlockFromGCPStrategyMap(strat)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	serviceAccountJson, err := getGCPCredentials(ctx, testingCtx.Client)
+	if err != nil {
+		t.Fatal("failed to retrieve gcp credentials %w", err)
```
Comment (cecobask) [inline test/functional/gcp_network_state.go:61]: AFAIK, using the `%w` verb can only be used with `fmt.Errorf()`.
The same problem can be seen in most of the `t.Fatal()` invocations here.
Something like that should do:
`t.Fatalf("<some error message>: %s", err)`

File: test/functional/gcp_network_state.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+	}
+	if expectedCidr != "" {
+		if cidr, err := strconv.ParseInt(expectedCidr, 10, 32); err != nil && address.GetPrefixLength() != int32(cidr) {
+			return nil, fmt.Errorf("address range cidr %d, does not match expected %s", address.GetPrefixLength(), expectedCidr)
+		}
+	}
+	return address, nil
+}
+
+func verifyGcpServiceConnection(ctx context.Context, client k8sclient.Client, network *computepb.Network, projectID string, opt option.ClientOption) (string, error) {
```
Comment (cecobask) [inline test/functional/gcp_network_state.go:137]:
```suggestion
func verifyGcpServiceConnection(ctx context.Context, network *computepb.Network, projectID string, opt option.ClientOption) (string, error) {
```

**Review by cecobask — CHANGES_REQUESTED** [review]
Left some comments which I'd like to see addressed.
I followed the verification steps and the E2E tests completed successfully:
```
[Prepare] Create namespace test-functional
[Prepare] Create serviceAccount cluster-admin-sa
[Prepare] Create ClusterRoleBinding for the service account
[integreatly-operator-test] Start test container
W0308 14:41:30.439681   29473 warnings.go:70] spec.template.spec.imagePullSecrets[0].name: invalid empty name ""
[integreatly-operator-test] Waiting for job to be started
[integreatly-operator-test] Pod found for job: integreatly-operator-test-77s9r
[integreatly-operator-test] Wait for test container to finish
[integreatly-operator-test] Pod=integreatly-operator-test-77s9r State=Pending
[integreatly-operator-test] Pod=integreatly-operator-test-77s9r Container=test ContainerStatus=Pending WatchStatus=ADDED
[integreatly-operator-test] Pod=integreatly-operator-test-77s9r PodStatus={Pending [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2023-03-08 14:41:30 +0000 GMT  } {Ready False 0001-01-01 00:00:00 +0000 UTC 2023-03-08 14:41:30 +0000 GMT ContainersNotReady containers with unready status: [test sidecar]} {ContainersReady False 0001-01-01 00:00:00 +0000 UTC 2023-03-08 14:41:30 +0000 GMT ContainersNotReady containers with unready status: [test sidecar]} {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2023-03-08 14:41:30 +0000 GMT  }]    10.11.129.3  [] 2023-03-08 14:41:30 +0000 GMT [] [{sidecar {&ContainerStateWaiting{Reason:ContainerCreating,Message:,} nil nil} {nil nil nil} false 0 busybox   0xc000b23c9e} {test {&ContainerStateWaiting{Reason:ContainerCreating,Message:,} nil nil} {nil nil nil} false 0 quay.io/tdimov0/integreatly-operator-test-harness:latest   0xc000b23cd4}] BestEffort []}
[integreatly-operator-test] Pod=integreatly-operator-test-77s9r State=Running
[integreatly-operator-test] Pod=integreatly-operator-test-77s9r Container=test ContainerStatus=Running
[integreatly-operator-test] Pod=integreatly-operator-test-77s9r State=Running
[integreatly-operator-test] Pod=integreatly-operator-test-77s9r Container=test ContainerStatus=Terminated
[integreatly-operator-test] Tests completed. Exit code = 0
[integreatly-operator-test] Save test pod status
[integreatly-operator-test] Download test results
WARNING: rsync command not found in path. Please use your package manager to install it.
./junit-integreatly-operator.xml
[integreatly-operator-test] Download test container logs
[integreatly-operator-test] Delete test job
W0308 15:56:21.250178   29473 warnings.go:70] child pods are preserved by default when jobs are deleted; set propagationPolicy=Background to remove them or set propagationPolicy=Orphan to suppress this warning
[integreatly-operator-test] Test container finished successfully
[Reporting] Tests completed. Results can be found in test-results/2023-03-08-02-41-29
```

File: test/functional/common_shared_functions.go [hunk 3]
```diff
@@ -81,6 +86,38 @@ func GetRedisInstancesIDs(ctx context.Context, client client.Client, rhmi *integ
...
+		// return resource id
+		resourceID, err := getCROAnnotation(postgres)
+		if err != nil {
+			foundErrors = append(foundErrors, fmt.Sprintf("\n%s postgres cr does not contain a resource id annotation: %v", r, err))
+		}
+		// populate the array
+		foundResourceIDs = append(foundResourceIDs, resourceID)
+	}
+	return foundResourceIDs, foundErrors
+}
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply test/functional/common_shared_functions.go:119]: I have updated this here and also in the pre-existing `GetRedisInstancesIDs` function.

File: test/functional/gcp_memorystore_redis_instance_exist.go [hunk 4]
```diff
@@ -99,3 +81,9 @@ func verifyRedisInstances(redisInstanceName string, databaseInstanceList []strin
 	}
 	return false
 }
+
+func verifyRedisInstanceConfig(instance *redispb.Instance) bool {
+	return memorystoreRedisLabelContains(instance, managedLabelKey, managedLabelValue) &&
+		instance.RedisVersion == "REDIS_6_X" && instance.State == redispb.Instance_READY &&
+		instance.Tier == redispb.Instance_STANDARD_HA && instance.MaintenancePolicy != nil
+}
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply test/functional/gcp_memorystore_redis_instance_exist.go:89]: Updating this here and in CRO.

File: test/functional/gcp_postgressql_instances_exist.go [hunk 5]
```diff
@@ -2,66 +2,51 @@ package functional
...
-	for _, v := range databaseInstanceList {
-		if v == pSqlInstanceName {
-			return true
-		}
-	}
-	return false
+func verifyPostgresSQLInstanceConfig(instance *sqladmin.DatabaseInstance) bool {
+	return postgresSqlLabelContains(instance, managedLabelKey, managedLabelValue) &&
+		instance.Settings.DeletionProtectionEnabled && instance.DatabaseVersion == "POSTGRES_13" && instance.Settings.AvailabilityType == "REGIONAL"
 }
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply test/functional/gcp_postgressql_instances_exist.go:52]: Updating this here and in CRO.

File: test/functional/gcp_shared_functions.go [hunk 6]
```diff
@@ -114,3 +72,35 @@ func parseSubnetUrl(subnetUrl string) (string, string, error) {
...
+
+func memorystoreRedisLabelContains(instance *redispb.Instance, key, value string) bool {
+	labels := instance.Labels
+	for k, v := range labels {
+		if k == key && v == value {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply test/functional/gcp_shared_functions.go:106]: Updated, thank you

File: test/functional/gcp_network_state.go (hunk 1 above)
Comment (Fiona-Waters) [reply test/functional/gcp_network_state.go:61]: Updated

File: test/functional/gcp_network_state.go (hunk 2 above)
Comment (Fiona-Waters) [reply test/functional/gcp_network_state.go:137]: Updated

File: test/functional/gcp_memorystore_redis_instance_exist.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (adam-cattermole) [reply test/functional/gcp_memorystore_redis_instance_exist.go]: This currently matches the AWS equivalent but there was talk recently about using the version from the CR instead. I don't have a strong preference to whether we use the default in CRO code or the version from CR but if we're going to change it we should change AWS to match in my opinion.

[the remark this answers is not in the document]
Comment (cecobask) [reply test/functional/gcp_memorystore_redis_instance_exist.go]: @adam-cattermole, getting the version from the CR sounds like a more elegant and error prone solution!
@Fiona-Waters, would you mind updating the tests to retrieve the Redis / Postgres version from their CR?

[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply test/functional/gcp_memorystore_redis_instance_exist.go]: Yes no problem.

File: test/functional/common_shared_functions.go [hunk 7]
```diff
@@ -54,31 +59,72 @@ Each resource provisioned contains an annotation with the resource ID
...
+			foundErrors = append(foundErrors, fmt.Errorf("\n%s postgres cr does not contain a resource id annotation: %v", r, err))
+		}
+
+		if postgres.Status.Version != "" {
+			postgresDatabaseVersion = postgres.Status.Version
+		}
+		foundData[resourceID] = postgresDatabaseVersion
+	}
+	return foundData, foundErrors
 }
```
Comment (cecobask) [inline test/functional/common_shared_functions.go:128]:
```suggestion
// GetPostgresInstanceData Each resource provisioned contains an annotation with the resource ID.
// This function iterates over a list of expected Postgres CRs and returns a map
// with key that represents the resource ID and value - its database version.
func GetPostgresInstanceData(ctx context.Context, client client.Client, rhmi *integreatlyv1alpha1.RHMI) (map[string]string, []error) {
	var foundErrors []error
	var instanceData map[string]string
	for _, pgName := range getExpectedPostgres(rhmi.Spec.Type, rhmi.Name) {
		postgres := &crov1.Postgres{}
		if err := client.Get(ctx, types.NamespacedName{Namespace: common.RHOAMOperatorNamespace, Name: pgName}, postgres); err != nil {
			foundErrors = append(foundErrors, fmt.Errorf("failed to find %s postgres cr: %w", pgName, err))
		}
		if postgres.Status.Phase != croTypes.PhaseComplete {
			foundErrors = append(foundErrors, fmt.Errorf("found postgres %q with phase %q but expected %q; message: %s", pgName, postgres.Status.Phase, croTypes.PhaseComplete, postgres.Status.Message))
		}
		resourceID, err := getCROAnnotation(postgres)
		if err != nil {
			foundErrors = append(foundErrors, fmt.Errorf("postgres cr %q does not contain a resource id annotation: %w", pgName, err))
		}
		instanceData[resourceID] = postgres.Status.Version
	}
	return instanceData, foundErrors
}
```

File: test/functional/common_shared_functions.go [hunk 8]
```diff
@@ -54,31 +59,72 @@ Each resource provisioned contains an annotation with the resource ID
...
-	var foundErrors []string
-	var foundResourceIDs []string
+func GetRedisInstanceData(ctx context.Context, client client.Client, rhmi *integreatlyv1alpha1.RHMI) (map[string]string, []error) {
+	var foundErrors []error
+	var foundData map[string]string
+	var redisVersion string
 
 	expectedRedis := getExpectedRedis(rhmi.Spec.Type, rhmi.Name)
 
 	for _, r := range expectedRedis {
```
Comment (cecobask) [inline test/functional/common_shared_functions.go:69]: GitHub won't allow me to make suggestions for deleted lines but this is what I think the function should look like:
```suggestion
// GetRedisInstanceData Each resource provisioned contains an annotation with the resource ID.
// This function iterates over a list of expected Redis CRs and returns a map
// with key that represents the resource ID and value - its database version.
func GetRedisInstanceData(ctx context.Context, client client.Client, rhmi *integreatlyv1alpha1.RHMI) (map[string]string, []error) {
	var foundErrors []error
	var instanceData map[string]string
	for _, redisName := range getExpectedRedis(rhmi.Spec.Type, rhmi.Name) {
		redis := &crov1.Redis{}
		if err := client.Get(ctx, types.NamespacedName{Namespace: common.RHOAMOperatorNamespace, Name: redisName}, redis); err != nil {
			foundErrors = append(foundErrors, fmt.Errorf("failed to find %s redis cr: %w", redisName, err))
		}
		if redis.Status.Phase != croTypes.PhaseComplete {
			foundErrors = append(foundErrors, fmt.Errorf("found redis %q with phase %q but expected %q; message: %s", redisName, redis.Status.Phase, croTypes.PhaseComplete, redis.Status.Message))
		}
		resourceID, err := getCROAnnotation(redis)
		if err != nil {
			foundErrors = append(foundErrors, fmt.Errorf("redis cr %q does not contain a resource id annotation: %w", redisName, err))
		}
		instanceData[resourceID] = redis.Status.Version
	}
	return instanceData, foundErrors
}
```

**Review by cecobask — CHANGES_REQUESTED** [review]

**Comment by adam-cattermole** [comment]
Going to see if we have any luck running these on the PR today

/test all

**Comment by cecobask** [comment]
The requested changes have been addressed.
Most commits were squashed, assuming everything is fine there and approving 👍🏻 
Thanks, @Fiona-Waters!

/lgtm

**Pull request merged**